### ISSUE TITLE
PUT in REST API does not upsert object with custom ID

### ISSUE THREAD

**Issue by ArkeshGKalathiya** [parse-community/parse-server#7139]
### New Feature / Enhancement Checklist

- [x] I am not disclosing a [vulnerability](https://github.com/parse-community/parse-server/blob/master/SECURITY.md).
- [x] I am not just asking a [question](https://github.com/parse-community/.github/blob/master/SUPPORT.md).
- [x] I have searched through [existing issues](https://github.com/parse-community/parse-server/issues?q=is%3Aissue).

### Current Limitation
When integrating third party apis with server,  most of them already provide the unique ids for the returned data. So right now if I want to save object with custom id it requires restful api call from node environment, and to make restful api call ( batch request ), we need to check with the server if object with custom id exist or not. If object exist then batch array requires PUT otherwise it requires POST. So it is kind of little overhead and extra management of the code. And also in fast updating environment it can cause errors ( where multiple node in same cluster checks with the backend if certain object id exist, it can cause race conditions ). so It would be nice to have support for saving objects in node environment without rest apis, and maybe upsert option when saving parse object.

### Feature / Enhancement Description
So It would be great to have feature where parse server sdk supports saving object based on custom object ids ( it doesn't support at the moment in cloud code and server sdk, it requires rest api cal in node environment to save such objects with custom object id ) and one flag with something like Parse.Object.save({},{upsert:true}), which updates existing documents if object id is already there in database.

### Alternatives / Workarounds
Making redundant requests to check if object exist in server

**Comment by mtrezza** [comment]
Related: https://github.com/parse-community/parse-server/issues/4170

**Comment by ArkeshGKalathiya** [comment]
Hi @mtrezza, Thanks for the response.

I have confirmed that PUT request for non existent object is not working, it gives error.

https://docs.parseplatform.org/rest/guide/#updating-objects

**Comment by mtrezza** [comment]
Thanks for confirming, can you please write a test case for this?

**Comment by ArkeshGKalathiya** [comment]
Can you please give me suggestion about where I can write test case? Do I have to pull parse-server repo and put test case there? or separate code would be sufficient?

**Comment by mtrezza** [comment]
Yes, you would
- fork Parse Server
- clone it locally
- create a new branch
- add a test to an already existing test file, such as https://github.com/parse-community/parse-server/blob/master/spec/ParseServerRESTController.spec.js (there may be another one that is more focused on schema changes)
- push the branch to your fork on github
- submit a PR to the Parse Server repo with your branch 

You can find more details in the Contribution guide. Don't get scared, the guide needs some rework and we're working on it. If you need any guidance please feel free to ask.

**Comment by mtrezza** [comment]
@ArkeshGKalathiya Do you need any help for creating a test case?

**Comment by ArkeshGKalathiya** [comment]
@mtrezza No, I was just busy with other stuff. I will soon add test case and submit PR.

**Comment by dplewis** [comment]
@ArkeshGKalathiya I added customID to the JS SDK. Feel free to provide feedback

https://github.com/parse-community/Parse-SDK-JS/pull/1309

**Comment by mstniy** [comment]
> Hi @mtrezza, Thanks for the response.
> 
> I have confirmed that PUT request for non existent object is not working, it gives error.
> 
> https://docs.parseplatform.org/rest/guide/#updating-objects

This issue with PUT requests still persists. I believe this issue should be re-opened.

**Comment by mtrezza** [comment]
I'm not sure why this was closed, @dplewis?

I think we can reopen if we have a failing test case?

**Comment by mstniy** [comment]
I don't think there is a test case that covers this. `ParseAPI.spec.js` has a few test for `PUT`, but none of them cover the case of using custom object ids to create a new object.

**Comment by mtrezza** [comment]
@mstniy Would you want to add a test case?

**Comment by mstniy** [comment]
Sure thing.

My understanding is that `POST` should create a new object. For requests with a custom object id, if the object already exists, it should fail. Whereas `PUT` should basically be an upsert. Please let me know if I am mistaken.

Should the PR fix the issue with `PUT` as well as introducing corresponding regression tests?

**Comment by mtrezza** [comment]
> What should a pure (non-upsert) update look like?

A `PATCH` request. Unlike `PUT` which is always a full representation of an object, a `PATCH` can is a partial representation and contains instructions on how to modify a resource.

That is the HTTP standard. That being said, if Parse is currently violating that standard, we'd have to look into what a correction implies in terms of breaking changes.

**Comment by mstniy** [comment]
It seems Parse is indeed violating the standard. From [the docs about PUT requests](https://docs.parseplatform.org/rest/guide/#updating-objects):

> Any keys you don’t specify will remain unchanged, so you can update just a subset of the object’s data. 

So the current `PUT` is actually a `PATCH`: It is a strict update, not an upsert, and it takes as parameter a partial representation.
How about using query strings? For example, we can enable upsert semantics only on `PUT` requests of the form `/classes/Foo/test_id?upsert=true`

**Comment by mtrezza** [comment]
> Any keys you don’t specify will remain unchanged, so you can update just a subset of the object’s data.

That sounds like a violation of http/1.1. A `PUT` must include a full object, not a partial object.

> So the current PUT is actually a PATCH

What happens when making a `PUT` request for an object that doesn't exist?

**Comment by mstniy** [comment]
Quoting [the specification you have mentioned](https://www.w3.org/Protocols/rfc2616/rfc2616-sec9.html#sec9.6):

> If the Request-URI does not point to an existing resource, and that URI is capable of being defined as a new resource by the requesting user agent, the origin server can create the resource with that URI. If a new resource is created, the origin server MUST inform the user agent via the 201 (Created) response. 

Currently, Parse returns an OBJECT_NOT_FOUND

**Comment by mtrezza** [comment]
You get that response for PUT or PATCH?

**Comment by mstniy** [comment]
For PUT

I don't think Parse uses PATCH?

**Comment by mtrezza** [comment]
So the main issue is that a PUT does not create an object if it doesn't exists yet, ergo its not an upsert.

**Comment by mstniy** [comment]
I agree
But I believe fixing it per the specification would likely break existing code

**Comment by mtrezza** [comment]
I suppose fixing (a) would be a breaking change, but possibly not have wide implications for Parse SDKs. It would only be a breaking change for mechanisms that are pending on a PUT failing for non-existent objects. Making a PUT for a non-existent object is probably unintended in most cases anyway.

Fixing (b) would likely affect all Parse SDKs, and may not be worth the effort for now.

**Comment by mstniy** [comment]
Introducing upserts as a query string option sounds like a good compromise.

**Comment by mtrezza** [comment]
Not sure about that, shouldn't we rather go i to the direction of correcting the behavior instead of leaving the incorrect behavior in place and adding an option?

**Comment by mstniy** [comment]
I would be happy to do that, as our use case wouldn't be negatively affected.
But it will likely be a breaking change for certain users.

**Comment by mtrezza** [comment]
A breaking change that is actually a correction is generally acceptable. 

I am just not sure what that change implies.
- Do we need to make changes in Parse SDKs as well?
- Are any other mechanisms in Parse Server affected?
- In which scenarios would that be breaking change for existing deployments; any examples?

**Comment by mstniy** [comment]
Here's my take:

- We rename the existing PUT operation to PATCH.
- We make the PUT use upsert and take as parameter full objects.
- We switch ParseObject.save() from doing a PUT to a PATCH.
- We change ParseObject.save() to do a PUT if the object has an id, but no createdAt.
    - This alters its behaviour. Current PUT fails for non-existing objects, but the new PUT will create them.
- We need a new method for PUTs, maybe ParseObject.put()?

I am not sure as to whether any mechanism would be affected.

**Comment by mstniy** [comment]
Here's my take:

    We rename the existing PUT operation to PATCH.
    We make the PUT use upsert and take as parameter full objects.
    We switch ParseObject.save() from doing a PUT to a PATCH.
    We need a new method for PUTs, maybe ParseObject.put()?

I am not sure as to whether any mechanism would be affected.
And of course, this could break any code base that uses the REST api directly. The fix would be to change from PUT to PATCH.

**Comment by mtrezza** [comment]
It seems what you are suggesting is to become fully spec compliant.

If that was easily achievable without major implications I would agree. If there are major implications, we may need to be a bit more pragmatic. Parse Server works after all, even if it is not fully spec compliant.

We could:
- add new PATCH route for partial update functionality -> no breaking change, evaluating of Parse Server internals needed, for example this requires changes in Idempotency feature
- add upsert functionality to existing PUT while keeping partial update functionality -> breaking change, evaluation of all Parse SDKs and Parse Server internals needed
- remove partial update functionality from existing PUT -> breaking change with possibly wide implications, across all Parse SDKs and custom code for most Parse Server deployments that use the REST API

**Comment by mtrezza** [comment]
Reopened because issue still seems to exist.

**Comment by cbaker6** [comment]
> Here's my take:
> 
> * We rename the existing PUT operation to PATCH.
> * We make the PUT use upsert and take as parameter full objects.
> * We switch ParseObject.save() from doing a PUT to a PATCH.
> * We need a new method for PUTs, maybe ParseObject.put()?
> 
> I am not sure as to whether any mechanism would be affected.
> And of course, this could break any code base that uses the REST api directly. The fix would be to change from PUT to PATCH.

@mstniy suggested steps seems reasonable to me as they would hopefully yield the least of amount of errors added to the server (and salvage all of the current `PUT` testcases) as you both mentioned that the current parse-server's `PUT` acts as `PATCH`.

The changes on the client SDKs will most likely follow the same process and should be easier to do (thinking from the Swift SDK and probably some of the others, this would be straight forward).  To the point of:

> We need a new method for PUTs, maybe ParseObject.put()?

Will something like `.replace()` work for put? This way on the client side, `save` will do `POST/PATCH (created/updated)` and if the user really wants to replace an object, they use `.replace()`.

Maybe when these changes occur, we can take advantage of the `X-Parse-Client-Version` header (I believe on the JS SDK is currently using this, but we can add the header to all `PUT` calls on all "patched" clients). We basically modify this on the server https://github.com/parse-community/parse-server/blob/9ea355b4635226ae4da17c8cc5fb0321e3fdec5e/src/ClientSDK.js#L3-L22

When a client's version is compatible, use the correct version of `PUT`. If the client version is less, or no header provided (allow backwards compatibility with older client versions) use `PATCH` on the server instead of `PUT`. This will allow the "patched" server to be able to deal with any client version.

Of course, if someone uses a "patched" client SDK with an non "patched" server they will have unexpected behavior.

**Comment by cbaker6** [comment]
@mtrezza bringing your attention back here as it seems like an important discussion. I’m not sure if a PR was ever started by @mstniy

**Comment by mstniy** [comment]
No, I haven't created a PR yet

**Comment by mtrezza** [comment]
Thanks, I think we agree so far that  this requires a significant change affecting all SDKs, if we want to solve this sustainably.

Deprecating a feature across all SDKs is a novelty in that context. The question is how to do it in a way that is:

- a) technically sustainable
- b) transparent for the developer dealing with multiple versions of SDKs
- c) according to our deprecation policy to give the developer time to adjust while still being able to upgrade to new versions

Luckily we have some tools:

- using the `X-Parse-Client-Version` - adding a compatibility table for Parse SDKs <-> Parse Server, this is a missing piece already now
- the Parse Server deprecator

> This way on the client side, save will do POST/PATCH (created/updated) and if the user really wants to replace an object, they use .replace().

How do you mean? Replacing a full object would be a `PATCH`. There is no HTTP method to replace a full object.

**Comment by cbaker6** [comment]
> How do you mean? Replacing a full object would be a PATCH. There is no HTTP method to replace a full object.

I think what I meant here is `.replace()` would call the “new“ `PUT` after `PATCH` is added.

**Comment by cbaker6** [comment]
> c) according to our deprecation policy to give the developer time to adjust while still being able to upgrade to new versions

I would imagine the developers who will run into the most issues are the ones who don’t use the Client SDKs and instead use REST or Graph directly, but we can assume those who to send the `X-Parse-Client-Version` are using the deprecated way

**Comment by cbaker6** [comment]
> Do we already agree on which approach to take? I think we'd need to compare implications.

Following @mtrezza’s approach seems right to deprecate in a step-by-step way. My question here, is this process needed because of the tools already available? Meaning following the process @mstniy mentioned and I commented on [here](https://github.com/parse-community/parse-server/issues/7139#issuecomment-842357245) may fix this now without causing issues (I’m saying this without seeing the CI of course) because it will automatically support the current way and the new way

**Comment by mtrezza** [comment]
Do we need a dedicated method for this?

If an object is saved with an `objectId`, then `PUT` will create a new object if it doesn't exist, or update all fields and remove non-specified fields.

**Comment by mtrezza** [comment]
Maybe I'm missing something, but what if we just add an option to `PUT` to make it behave like a `PATCH`? The SDKs can offer an `obj.save({ upsert: true });` option  or a dedicated `obj.upsert();` method whenever they are ready.

We wouldn't be become any more or less spec compliant, but this wouldn't break anything.

We can optionally add a `PATCH` route to Parse Server, to become a bit more compliant.

**Comment by cbaker6** [comment]
Some things to think about when it comes to server responses for ParseObjects. Currently a `PUT` returns `updatedAt` to the client, while a `POST` returns `createdAt`.

The proposed changes seem to leave `POST` unchanged, `PATCH` returns `updatedAt`, and `POST` returning either `createdAt` or `updatedAt`? For the new `POST`, the logic will need to be handled in the client SDKs.

**Comment by mtrezza** [comment]
TL;DR:




Let's try to break this down...

There are two separate issues:

- a) Fix the issue of not offering a way to upsert an object. Upsert means to create a new object if no `objectId` is set or the set `objectId` does not exist, or replace an entire object if the `objectId` exists.
- b) Fix the issue of Parse Server not being http spec compliant because it doesn't have a `PATCH` route for partial object updates, but the `PUT` route currently behaves as would be expected from a `PATCH` route.

These two issues can looked at individually, because:

- to fix (a) it is not is not required to fix (b), but fixing (b) will also fix (a)
- while (a) is an actually issue that impedes development with Parse Server, (b) is a mere formality that does not (as much)
- (b) requires a change that breaks the backward compatibility of old SDKS with new Parse Server with old SDKs; this happens when Parse Server starts to treat a `PUT` request as a spec compliant `PUT` (upsert) instead of as currently a spec compliant `PATCH` (partial update).
  - old SDK <-> old Parse Server: no break
  - old SDK <-> new Parse Server: no break; Parse Server can use `X-Parse-Client-Version` to decide whether to treat `PUT` request as spec compliant `PUT` or `PATCH`
  - new SDK <-> new Parse Server: no break; same as old SDK <-> new Parse Server
  - ❌ new SDK <-> old Parse Server: break; Parse Server treats `PUT` request as spec compliant `PUT` and instead of partially updating the object, it is replaced entirely

This means that (b) will pose the challenge to developers that if they want to upgrade to the new SDK, they will have to upgrade to new Parse Server as well. They cannot only upgrade the SDK without upgrading Parse Server. Upgrading Parse Server is currently a major obstacle for many deployments, because Parse Server 5.0 will bring many changes, some of them breaking. Many deployments seem to run on old versions of Parse Server and there is a certain "upgrade fatigue" because in the past we had no deprecation policy, sudden breaking changes, infrequent and non-matured releases, no LTS, etc. This situation is expected to improve dramatically from Parse Server 5.0 onwards and over the course of 2022/23.

This means we are not in a good place at the moment to fix (b) by introducing a breaking change that will be a burden for many developers. It would also be an unnecessarily high burden, because (b) is merely a formality. So I suggest to fix (b) when we can expect a majority of developers having jumped onto the release train because they know that keeping their deployments up-to-date is almost effortless. I expect that to be sometime in 2023/24.

If we solely focus on (a), an approach could be:

1. Create PR to Parse Server to add `upsert` option to `PUT` request to tell Parse Server whether to treat the `PUT` like a spec compliant `PUT` or `PATCH`. Default should be `PATCH`, which is the status quo.
1. Create PR to add a compatibility table `Parse Server <-> Parse SDKs` to Parse Server. This is currently a missing instrument, regardless of this issue, but in fixing this issue it becomes even more important to create transparency for the developer because not every SDK will be compatible with every version of Parse Server.
1. Create PR to SDK can offer an obj.save({ upsert: true }); option or a dedicated obj.upsert(); method. This fixes (a) because it gives the developer a way to request a spec compliant `PUT`, which is currently not possible. I suggest to provide a dedicated obj.upsert(); method, because we can easily show a deprecation warning at compile time, whereas a `save` option would only probably only be evaluable during runtime or on the server side.
3. Developers can upgrade to new SDK, knowing that they will also have to upgrade to new Parse Server *if* they want to use the `upsert` option. All other developers who do not use the`upsert` option (that is the majority I suspect) can continue to upgrade only the SDK without having to upgrade Parse Server.

To prepare for the breaking change in 2023/24:
1. Create PR to add a `PATCH` route to Parse Server. This alone makes Parse Server a bit more http spec compliant, because currently is has no `PATCH` route at all, even if `PUT` behaves like `PATCH`. This PR can be done anytime, as it only adds a route.

Sometime in 2023/24 (Parse Server 6/7):
1. Create PR for SDK to use `PATCH` for partial object update, `PUT` for entire object replacement if exists or object creation if not exists. This creates an incompatibility with old Parse Server versions. This should be done once we assume that we have overcome the current "upgrade fatigue".

**Comment by cbaker6** [comment]
In attempt to have the Swift SDK as close to spec compliant as possible while still making it work with the current Parse Server, I made changes in https://github.com/parse-community/Parse-Swift/pull/299. 

Just further confirming that when I remove the [guard on objectId](https://github.com/parse-community/Parse-Swift/blob/ebaaf12b21f73ceda047928d4401b34b118d4e72/Sources/ParseSwift/API/API%2BCommand.swift#L415-L418) when attempting to `replace()`(PUT/upsert), I get the following when testing on `parse-server:5.0.0-beta.4`:

- `remove()` with no `objectId`: I receive a `404` with `Cannot PUT /1/classes/GameScore`. 
- `remove()` with a new `objectId`: ` error: Object not found. {"code":101,"stack":"Error: Object not found.\n    at /parse-server/lib/Controllers/DatabaseController.js:581:17\n    at processTicksAndRejections (internal/process/task_queues.js:95:5)"}`

The same results occur no matter the server setting, `allowCustomObjectId = true/false`.

I've created a [branch](https://github.com/cbaker6/Parse-Swift/tree/upsert) of ParseSwift for testing upsert in [Swift Playgrounds](https://github.com/cbaker6/Parse-Swift/blob/e7d8e76d324e26ad5105ee149a77c20e53d41652/ParseSwift.playground/Pages/1%20-%20Your%20first%20Object.xcplaygroundpage/Contents.swift#L81-L123) whenever development on the server-side for upsert and `update()`(PATCH) begins.

**Comment by ArkeshGKalathiya** [comment]
Any update? How can I help on this?

**Comment by mtrezza** [comment]
@ArkeshGKalathiya You could read through this thread and suggest a PR.
